README.md
# hmsr

<!-- badges: start -->

[![R-CMD-check](https://github.com/WojtAcht/hms/workflows/R-CMD-check/badge.svg)](https://github.com/WojtAcht/hms/actions)

<!-- badges: end -->

The HMS (Hierarchic Memetic Strategy) is a composite global optimization strategy consisting of a multi-population evolutionary strategy and some auxiliary methods. The HMS makes use of a dynamically-evolving data structure that provides an organization among the component populations. It is a tree with a fixed maximal height and variable internal node degree. Each component population is governed by a particular evolutionary engine. This package provides a simple R implementation with examples of using different genetic algorithms as the population engines.

### Relevant literature

-   J. Sawicki, M. Łoś, M. Smołka, R. Schaefer. Understanding measure-driven algorithms solving irreversibly ill-conditioned problems. Natural Computing 21:289-315, 2022. doi: [10.1007/s11047-020-09836-w](https://doi.org/10.1007/s11047-020-09836-w)

This package does not provide an implementation of a Simple Genetic Algorithm to be used in demes. However the default configuration uses SGA implementation from [GA](https://github.com/luca-scr/GA) package available on CRAN.

## Installation

You can install the released version of `hmsr` from CRAN:

``` r
install.packages("hmsr")
```

or the development version from this repository:

``` r
install.packages("devtools")
library("devtools")
devtools::install_github("WojtAcht/hms")
```

## Usage

### Default configuration

To run the HMS with a default configuration the only arguments that have to be provided are the bounds of the domain: `lower` and `upper`

``` r
library(hmsr)
f <- function(x)  (x^2+x)*cos(x)
HMS <- hms(fitness = f, lower = -10, upper = 10)
HMS@best_solution
#> [1] 6.560544
```

### Custom configuration

There are number of parameters used for strategy configuration. The sample below runs HMS on the Eggholder function - a benchmark for optimization problems.

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/e7/Eggholder_function.pdf/page1-1200px-Eggholder_function.pdf.jpg" width="400"/>

``` r
library(smoof)
Eggholder <- smoof::makeEggholderFunction()
lower <- rep(-512, 2)
upper <- rep(512, 2)
sigma <- list(rep(200, 2), rep(100, 2), rep(50, 2))
ga_config <- list(
    list(
        pmutation = 0.4,
        mutation = rtnorm_mutation(lower, upper, sigma[[1]])
    ),
    list(
        pmutation = 0.2,
        mutation = rtnorm_mutation(lower, upper, sigma[[2]])
    ),
    list(
        pmutation = 0.2,
        mutation = rtnorm_mutation(lower, upper, sigma[[3]])
    )
)
HMS <- hms(
    fitness = Eggholder,
    minimize = TRUE,
    tree_height = 3,
    lower = lower,
    upper = upper,
    run_metaepoch = ga_metaepoch(ga_config),
    population_sizes = c(50, 30, 15),
    sigma = sigma,
    gsc = gsc_max_fitness_evaluations(20000),
    sc = sc_max_metric(euclidean_distance, c(40, 20, 10)),
    lsc = lsc_metaepochs_without_improvement(15),
    monitor_level = "none",
    with_gradient_method = TRUE
)
```

To see the structure of demes tree at the end of the evaluation, use `printTree` function

``` r
printTree(HMS)
#> f(476.52, 431.37) = 943.07 evaluations: 7675 A
#> ├-- spr: (507.90, 402.59); f(480.26, 430.74) = 956.02 evaluations: 1955 A
#> |   └-- spr: (507.49, 400.58); f(482.35, 432.88) = 956.92 evaluations: 879 A
#> ├-- spr: (479.43, 436.05); f(478.82, 429.39) = 954.55 evaluations: 2341 A
#> |   └-- spr: (480.59, 430.58); f(482.35, 432.88) = 956.92 evaluations: 1075 A
#> └-- spr: (365.96, 500.10); f(507.31, 401.35) = 938.19 evaluations: 4293 A
#>     └-- spr: (441.83, 454.04); ***f(512.00, 404.23) = 959.64*** evaluations: 2000 A
```

To display a plot showing how best solution has changed during the evaluation use `plot(HMS)`.


NEWS.md
# hmsr 1.0.1
* `sc_max_metric()` uses centroids instead of sprouts (#54).
* fix M1 mac failing tests.


R/comparisons.R
hms_mutation_comparison <- function(dimensions, evaluations) {
  seeds <- 1:30

  fitness <- function(x) { -1 * Griewank(x) }
  lower <- rep(-600, dimensions)
  upper <- rep(600, dimensions)
  sigma <- list(rep(200, dimensions), rep(100, dimensions), rep(50, dimensions))

  run_hms <- function(with_ga_mutation, seed) {
    print(match.call())
    ga_config <- list(
      list(
        pmutation = 0.4
      ),
      list(
        pmutation = 0.2
      ),
      list(
        pmutation = 0.2
      )
    )
    norm_ga_config <- list(
      list(
        pmutation = 0.4,
        mutation = rtnorm_mutation(lower, upper, sigma[[1]])
      ),
      list(
        pmutation = 0.2,
        mutation = rtnorm_mutation(lower, upper, sigma[[2]])
      ),
      list(
        pmutation = 0.2,
        mutation = rtnorm_mutation(lower, upper, sigma[[3]])
      )
    )

    set.seed(seed)

    result <- hms(
      fitness = fitness,
      tree_height = 3,
      lower = lower,
      upper = upper,
      run_metaepoch = ga_metaepoch(if(with_ga_mutation) ga_config else norm_ga_config),
      population_size_per_tree_level = c(50, 30, 15),
      sigma = sigma,
      global_stopping_condition =   global_stopping_condition_max_fitness_evaluations(evaluations),
      sprouting_condition = max_metric_sprouting_condition(euclidean_distance, c(40, 20, 10)),
      local_stopping_condition = local_stopping_condition_metaepochs_without_improvement(15),
      monitor_level = "none"
    )
    printTree(result)
    print(result@best_fitness)
    result@best_fitness
  }

  norm_mutation <- mapply(function(seed) { run_hms(FALSE, seed) }, seeds)
  ga_mutation <- mapply(function(seed) { run_hms(TRUE, seed) }, seeds)


  cat("norm:", mean(norm_mutation), "\n")
  cat("ga:", mean(ga_mutation), "\n")

  list(norm = norm_mutation, ga = ga_mutation)
}

hms_vs_ga <- function(evaluations) {
  seeds <- 1141:1171

  fitness <- function(x) { -1 * Eggholder(x) }
  lower <- rep(-512, 2)
  upper <- rep(512, 2)
  sigma <- list(rep(250, 2), rep(150, 2), rep(50, 2))

  run_hms <- function(seed) {
    ga_config <- list(
      list(
        pmutation = 0.5
      ),
      list(
        pmutation = 0.2,
        mutation = rtnorm_mutation(lower, upper, sigma[[2]])
      ),
      list(
        pmutation = 0.2,
        mutation = rtnorm_mutation(lower, upper, sigma[[3]])
      )
    )

    set.seed(seed)

    result <- hms(
      fitness = fitness,
      tree_height = 3,
      lower = lower,
      upper = upper,
      run_metaepoch = ga_metaepoch(ga_config),
      population_size_per_tree_level = c(50, 30, 15),
      sigma = sigma,
      global_stopping_condition =   global_stopping_condition_max_fitness_evaluations(evaluations),
      sprouting_condition = max_metric_sprouting_condition(euclidean_distance, c(45, 20, 10)),
      local_stopping_condition = local_stopping_condition_metaepochs_without_improvement(8),
      monitor_level = "none"
    )
    printTree(result)
    print(result@best_fitness)
    result@best_fitness
  }

  evals <- 0
  ga_f <- function(x) {
    evals <<- evals + 1

    if (evals > evaluations) {
      -Inf
    } else {
      fitness(x)
    }
  }

  run_ga <- function(seed) {
    set.seed(seed)
    evals <<- 0

    result <- GA::ga(type = "real-valued", fitness = ga_f, lower = lower, upper = upper, monitor = FALSE, maxiter = (evaluations / 150)*5, popSize = 150)
    print(result@fitnessValue)

    result@fitnessValue
  }

  hms <- mapply(function(seed) { run_hms(seed) }, seeds)
  ga <- mapply(function(seed) { run_ga(seed) }, seeds)


  cat("hms:", mean(hms), "\n")
  cat("ga:", mean(ga), "\n")

  list(hms = hms, ga = ga)
}

hms_vs_ga_schwefel_5d <- function(evaluations) {
  seeds <- 1141:1171

  fitness <- function(x) { -1 * Schwefel(x) }
  lower <- rep(-500, 5)
  upper <- rep(500, 5)
  sigma <- list(rep(300, 5), rep(150, 5), rep(50, 5))

  run_hms <- function(seed) {
    ga_config <- list(
      list(
        pmutation = 0.4
      ),
      list(
        pmutation = 0.2 #,
        #mutation = rtnorm_mutation(lower, upper, sigma[[2]])
      ),
      list(
        pmutation = 0.2,
        mutation = rtnorm_mutation(lower, upper, sigma[[3]])
      )
    )

    set.seed(seed)

    result <- hms(
      fitness = fitness,
      tree_height = 3,
      lower = lower,
      upper = upper,
      run_metaepoch = ga_metaepoch(ga_config),
      population_size_per_tree_level = c(50, 25, 10),
      sigma = sigma,
      global_stopping_condition =   global_stopping_condition_max_fitness_evaluations(evaluations),
      sprouting_condition = max_metric_sprouting_condition(euclidean_distance, c(55, 30, 18)),
      local_stopping_condition = local_stopping_condition_metaepochs_without_improvement(7),
      monitor_level = "none"
    )
    printTree(result)
    print(result@best_fitness)
    result@best_fitness
  }

  evals <- 0
  ga_f <- function(x) {
    evals <<- evals + 1

    if (evals > evaluations) {
      -Inf
    } else {
      fitness(x)
    }
  }

  run_ga <- function(seed) {
    set.seed(seed)
    evals <<- 0

    result <- GA::ga(type = "real-valued", fitness = ga_f, lower = lower, upper = upper, monitor = FALSE, maxiter = (evaluations / 150)*5, popSize = 150)
    print(result@fitnessValue)

    result@fitnessValue
  }

  hms <- mapply(function(seed) { run_hms(seed) }, seeds)
  ga <- mapply(function(seed) { run_ga(seed) }, seeds)


  cat("hms:", mean(hms), "\n")
  cat("ga:", mean(ga), "\n")

  list(hms = hms, ga = ga)
}

get_hms_vs_ga_res <- function() {
  a_2k = hms_vs_ga(2000)
  a_4k = hms_vs_ga(4000)
  a_6k = hms_vs_ga(6000)
  a_8k = hms_vs_ga(8000)
  a_10k = hms_vs_ga(10000)
  a_12k = hms_vs_ga(12000)
  a_14k = hms_vs_ga(14000)
  a_16k = hms_vs_ga(16000)

  list(ga_2k = a_2k$ga, hms_2k = a_2k$hms, ga_4k = a_4k$ga, hms_4k = a_4k$hms, ga_6k = a_6k$ga, hms_6k = a_6k$hms, ga_8k = a_8k$ga, hms_8k = a_8k$hms, ga_10k = a_10k$ga, hms_10k = a_10k$hms, ga_12k = a_12k$ga, hms_12k = a_12k$hms, ga_14k = a_14k$ga, hms_14k = a_14k$hms, ga_16k = a_16k$ga, hms_16k = a_16k$hms)
}

get_hms_vs_ga_schwefel_5d_res <- function() {
  a_5k = hms_vs_ga_schwefel_5d(5000)
  a_10k = hms_vs_ga_schwefel_5d(10000)
  a_15k = hms_vs_ga_schwefel_5d(15000)
  a_20k = hms_vs_ga_schwefel_5d(20000)
  a_25k = hms_vs_ga_schwefel_5d(25000)
  a_30k = hms_vs_ga_schwefel_5d(30000)
  #a_35k = hms_vs_ga_schwefel_5d(35000)
  #a_40k = hms_vs_ga_schwefel_5d(40000)

  list(ga_5k = a_5k$ga, hms_5k = a_5k$hms, ga_10k = a_10k$ga, hms_10k = a_10k$hms, ga_15k = a_15k$ga, hms_15k = a_15k$hms, ga_20k = a_20k$ga, hms_20k = a_20k$hms, ga_25k = a_25k$ga, hms_25k = a_25k$hms, ga_30k = a_30k$ga, hms_30k = a_30k$hms)#, ga_35k = a_35k$ga, hms_35k = a_35k$hms, ga_40k = a_40k$ga, hms_40k = a_40k$hms)
}

plot_hms_vs_ga <- function(res) {
  cols = tail(rainbow(3, s = 0.1), n=2)
  # , names = c("2k", "2k", "4k", "4k", "6k", "6k", "8k", "8k", "10k", "10k", "12k", "12k", "14k", "14k", "16k", "16k")
  boxplot(res, at=c(1,2,5,6,9,10,13,14,17,18,21,22), col=rep(cols, times=8), ylab="f(x)", xlab="Fitness evaluations")
  points(x = c(1,2,5,6,9,10,13,14,17,18,21,22),
         y = mapply(mean, res),
         col = rep(tail(rainbow(3), n=2), times=6),
         pch = 16)
  legend("topleft", fill = tail(rainbow(3), n=2), legend = c("GA", "HMS"), horiz = T)
}
